packages/ratelimiter/rate_limiter.go

```diff
@@ -0,0 +1,110 @@
+package ratelimiter
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ReneKroon/ttlcache/v2"
+	"github.com/cockroachdb/errors"
+	"github.com/iotaledger/hive.go/autopeering/peer"
+	"github.com/iotaledger/hive.go/events"
+	"github.com/iotaledger/hive.go/logger"
+	"github.com/paulbellamy/ratecounter"
+	"go.uber.org/atomic"
+)
+
+// RateLimit contains information about rate limit values such as time interval and the limit.
+type RateLimit struct {
+	Interval time.Duration
+	Limit    int
+}
+
+// String return a string representation of the RateLimit instance.
+func (rl RateLimit) String() string {
+	return fmt.Sprintf("%d per %s", rl.Limit, rl.Interval)
+}
+
+// PeerRateLimiter is an object to count activity of peers
+// and notify the subscribers in case the limit of activity is exceeded.
+type PeerRateLimiter struct {
+	interval     time.Duration
+	limit        *atomic.Int64
+	hitEvent     *events.Event
+	peersRecords *ttlcache.Cache
+	log          *logger.Logger
+}
+
+// NewPeerRateLimiter returns a new instance of the PeerRateLimiter object.
+func NewPeerRateLimiter(interval time.Duration, limit int, log *logger.Logger) (*PeerRateLimiter, error) {
+	records := ttlcache.NewCache()
+	records.SetLoaderFunction(func(_ string) (interface{}, time.Duration, error) {
+		record := &limiterRecord{counter: ratecounter.NewRateCounter(interval), limitHitReported: atomic.NewBool(false)}
```

Comment:
Isn't it a bit of an overkill to use `ttlcache` and `ratecounter`? If I understand correctly, the cache will expire every second anyway and the record recreated (if configured for 1 second interval). Now why can't we simply count up the entry for each peer in the cache and if it hits the limit for we trigger the event. Then there's no need for `ratecounter` and `limitHitReported` anymore. Also, `ttlcache` should be thread-safe and it can simply be a mapping to an int64 to count no?

I think this should work because we specify the limit per interval when we configure. Overall, this will make it a bit leaner and easier to understand imo.

Reply:
The cache doesn't expire if you touch it. So every time you count activity for the peer, its cache record TTL gets reset.
We could use an int counter for every peer and timestamp of the last reset. But it would be a different rate-limiting algorithm, less accurate.

Reply:
https://github.com/ReneKroon/ttlcache says:
> Note (issue #25): by default, due to historic reasons, the TTL will be reset on each cache hit and you need to explicitly configure the cache to use a TTL that will not get extended.

`SkipTTLExtensionOnHit(true)` should expire within the limit even when touching it.

Reply:
Yes, we can do this. What you are proposing is indeed simpler, but it would be a different rate-limiting algorithm, your version is called Fixed Window and the one using the rate counter is Sliding Window.
The fixed window is less accurate since it has a notion of hard reset of the timer, it's possible to generate twice more activity than Is set by the rate limiter. 
For example, we have a rate limit of 10 messages per second. With Fixed Window it's possible to spam 20 messages per second: 10 messages sent within 0.5 seconds before the time reset and 10 messages sent within 0.5 seconds after the time reset results in a total of 20 messages per second.

Reply:
Let's leave it with the sliding window approach but keep the other one in mind in case this is too inefficient.